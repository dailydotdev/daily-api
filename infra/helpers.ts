import * as pulumi from '@pulumi/pulumi';
import * as gcp from '@pulumi/gcp';
import * as inputs from '@pulumi/gcp/types/input';

function camelToUnderscore(key: string): string {
  const result = key.replace(/([A-Z])/g, ' $1');
  return result.split(' ').join('_').toUpperCase();
}

export const config = new pulumi.Config();
export const stack = pulumi.getStack();
export const location = gcp.config.region || 'us-central1';

export const infra = new pulumi.StackReference(`idoshamun/infra/${stack}`);

export function createEnvVarsFromSecret(
  prefix: string,
): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv>[] {
  const envVars = config.requireObject<Record<string, string>>('env');
  return Object.keys(envVars).map(
    (key): pulumi.Input<inputs.cloudrun.ServiceTemplateSpecContainerEnv> => {
      const secret = new gcp.secretmanager.Secret(`${prefix}-secret-${key}`, {
        secretId: `${prefix}-secret-${key}`,
        replication: { automatic: true },
      });

      const version = new gcp.secretmanager.SecretVersion(
        `${prefix}-sv-${key}`,
        {
          enabled: true,
          secret: secret.name,
          secretData: envVars[key],
        },
      );
      return {
        name: camelToUnderscore(key),
        value: pulumi
          .all([secret.secretId, version.id])
          .apply(
            ([name, version]) =>
              `gcp:///${name}/${version.split('/').reverse()[0]}`,
          ),
      };
    },
  );
}

export function serviceAccountToMember(
  serviceAccount: gcp.serviceaccount.Account,
): pulumi.Output<string> {
  return serviceAccount.email.apply((email) => `serviceAccount:${email}`);
}

export type IAMRole = { name: string; role: string };

export function addIAMRolesToServiceAccount(
  prefix: string,
  roles: IAMRole[],
  serviceAccount: gcp.serviceaccount.Account,
): void {
  const member = serviceAccountToMember(serviceAccount);
  roles.forEach(
    (role) =>
      new gcp.projects.IAMMember(`${prefix}-iam-${role.name}`, {
        role: role.role,
        member,
      }),
  );
}
