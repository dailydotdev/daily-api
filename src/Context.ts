import { Connection, EntitySchema, ObjectType, Repository } from 'typeorm';
import { FastifyRequest, Logger } from 'fastify';
import { RootSpan } from '@google-cloud/trace-agent/build/src/plugin-types';
import { GraphQLDatabaseLoader } from '@mando75/typeorm-graphql-loader';

export class Context {
  req: FastifyRequest;
  con: Connection;
  loader: GraphQLDatabaseLoader;

  constructor(req: FastifyRequest, con: Connection) {
    this.req = req;
    this.con = con;
    this.loader = new GraphQLDatabaseLoader(con);
  }

  get userId(): string | null {
    return this.req.userId;
  }

  get log(): Logger {
    return this.req.log;
  }

  get span(): RootSpan {
    return this.req.span;
  }

  getRepository<Entity>(
    target: ObjectType<Entity> | EntitySchema<Entity> | string,
  ): Repository<Entity> {
    return this.con.getRepository(target);
  }
}
