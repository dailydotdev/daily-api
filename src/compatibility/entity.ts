import { SourceRequest } from '../entity';
import { ChangeObject } from '../types';

export interface LegacySourceRequest {
  id: string;
  url: string;
  userId: string;
  userName?: string;
  userEmail?: string;
  approved?: boolean;
  closed?: boolean;
  pubId?: string;
  pubName?: string;
  pubImage?: string;
  pubTwitter?: string;
  pubRss?: string;
  createdAt: Date;
  reason?: string;
}

export const toLegacySourceRequest = (
  req: ChangeObject<SourceRequest> | SourceRequest,
): LegacySourceRequest => ({
  id: req.id,
  url: req.sourceUrl,
  userId: req.userId,
  userName: req.userName,
  userEmail: req.userEmail,
  approved: req.approved,
  closed: req.closed,
  pubId: req.sourceId,
  pubName: req.sourceName,
  pubImage: req.sourceImage,
  pubTwitter: req.sourceTwitter,
  pubRss: req.sourceFeed,
  createdAt: new Date(req.createdAt),
  reason: req.reason,
});
