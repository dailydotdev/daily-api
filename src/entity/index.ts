export * from './Notification';
export * from './Source';
export * from './SourceDisplay';
export * from './SourceFeed';
export * from './SourceRequest';
