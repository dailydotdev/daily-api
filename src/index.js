import Koa from 'koa';
import bodyParser from 'koa-bodyparser';
import KoaPinoLogger from 'koa-pino-logger';
import Router from 'koa-router';
import etag from 'koa-etag';


import config from './config';
// import compress from './middlewares/compress';
import errorHandler from './middlewares/errorHandler';
import logger from './logger';

import health from './routes/health';
import sources from './routes/sources';
import posts from './routes/posts';
import publications from './routes/publications';
import tweet from './routes/tweet';
import settings from './routes/settings';
import feeds from './routes/feeds';
import notifications from './routes/notifications';
import tags from './routes/tags';

const app = new Koa();

app.keys = [config.cookies.secret];

app.proxy = true;

// TODO: disabled due to performance issues
// app.use(compress());

app.use(bodyParser());
app.use(KoaPinoLogger({ logger, useLevel: 'debug' }));
app.use(errorHandler());
app.use(etag());

// Machine-to-machine authentication
app.use((ctx, next) => {
  if (ctx.request.get('authorization') === `Service ${config.accessSecret}`
    && ctx.request.get('user-id') && ctx.request.get('logged-in')) {
    // eslint-disable-next-line
    ctx.state = {
      user: {
        userId: ctx.request.get('user-id'),
      },
      service: true,
    };
  } else {
    delete ctx.request.headers['user-id'];
    delete ctx.request.headers['logged-in'];
  }
  return next();
});

const router = new Router({
  prefix: '/v1',
});

router.use(feeds.routes(), feeds.allowedMethods());
router.use(sources.routes(), sources.allowedMethods());
router.use(posts.routes(), posts.allowedMethods());
router.use(publications.routes(), publications.allowedMethods());
router.use(tweet.routes(), tweet.allowedMethods());
router.use(settings.routes(), settings.allowedMethods());
router.use(notifications.routes(), notifications.allowedMethods());
router.use(tags.routes(), tags.allowedMethods());

app.use(router.routes(), router.allowedMethods());
app.use(health.routes(), health.allowedMethods());

export default app;
