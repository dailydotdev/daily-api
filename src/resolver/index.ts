export * from './NotificationResolver';
export * from './SourceResolver';
export * from './SourceRequestResolver';
