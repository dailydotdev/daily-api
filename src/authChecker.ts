import { AuthChecker } from 'type-graphql';
import { Context } from './Context';
import { fetchUserRoles } from './users';

export enum Roles {
  Moderator = 'moderator',
}

export const authChecker: AuthChecker<Context, Roles> = async (
  { context },
  roles,
) => {
  if (!context.userId) {
    return false;
  }
  if (!roles.length) {
    return true;
  }

  const userRoles = await fetchUserRoles(context.userId);
  return roles.findIndex((r) => userRoles.indexOf(r) > -1) > -1;
};
