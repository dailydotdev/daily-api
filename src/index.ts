import 'reflect-metadata';
import fastify, {
  FastifyRequest,
  FastifyInstance,
  FastifyError,
} from 'fastify';
import fastifyRawBody from 'fastify-raw-body';
import helmet from '@fastify/helmet';
import cors from '@fastify/cors';
import mercurius, { MercuriusError } from 'mercurius';
import MercuriusGQLUpload from 'mercurius-upload';
import MercuriusCache from 'mercurius-cache';
import { NoSchemaIntrospectionCustomRule } from 'graphql';
// import fastifyWebsocket from '@fastify/websocket';

import './config';

import tracking from './tracking';
import auth from './auth';
import compatibility from './compatibility';
import routes from './routes';
import { Context } from './Context';
import { schema } from './graphql';
import createOrGetConnection from './db';
import { stringifyHealthCheck } from './common';
import { GraphQLError } from 'graphql';
import cookie, { FastifyCookieOptions } from '@fastify/cookie';
import { getSubscriptionSettings } from './subscription';
import { ioRedisPool } from './redis';
import { loadFeatures } from './growthbook';
import { runInRootSpan } from './telemetry/opentelemetry';
import { loggerConfig } from './logger';

type Mutable<Type> = {
  -readonly [Key in keyof Type]: Type[Key];
};

const userExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined => (ctx.userId ? `user:${ctx.userId}` : undefined);

const trackingExtendKey = (
  source: unknown,
  args: unknown,
  ctx: Context,
): string | undefined =>
  ctx.trackingId ? `tracking:${ctx.trackingId}` : undefined;

// readiness probe is set failureThreshold: 2, periodSeconds: 2 (4s) + small delay
const GRACEFUL_DELAY = 2 * 2 * 1000 + 5000;

export default async function app(
  contextFn?: (request: FastifyRequest) => Context,
): Promise<FastifyInstance> {
  let isTerminating = false;
  const isProd = process.env.NODE_ENV === 'production';
  const connection = await runInRootSpan(
    'createOrGetConnection',
    createOrGetConnection,
  );

  const app = fastify({
    logger: loggerConfig,
    disableRequestLogging: true,
    trustProxy: true,
  });

  app.log.info('loading features');
  await loadFeatures(app.log);

  const gracefulShutdown = () => {
    app.log.info('starting termination');
    isTerminating = true;
    setTimeout(async () => {
      await app.close();
      await connection.destroy();
      await ioRedisPool.end();
      process.exit();
    }, GRACEFUL_DELAY);
  };
  process.on('SIGINT', gracefulShutdown);
  process.on('SIGTERM', gracefulShutdown);

  app.register(helmet);
  app.register(cors, {
    origin: isProd ? /daily\.dev$/ : true,
    credentials: true,
  });
  app.register(cookie, {
    secret: process.env.COOKIES_KEY,
  }) as FastifyCookieOptions;
  app.register(auth, { secret: process.env.ACCESS_SECRET });
  app.register(tracking);
  app.register(fastifyRawBody, {
    field: 'rawBody',
    global: false,
    encoding: 'utf8',
    runFirst: true,
  });

  app.setErrorHandler((err, req, res) => {
    req.log.error({ err }, err.message);
    res.code(500).send({ statusCode: 500, error: 'Internal Server Error' });
  });

  app.get('/health', (req, res) => {
    res.type('application/health+json');
    if (isTerminating) {
      res.status(500).send(stringifyHealthCheck({ status: 'terminating' }));
    } else {
      res.send(stringifyHealthCheck({ status: 'ok' }));
    }
  });

  app.get('/liveness', (req, res) => {
    res.type('application/health+json');
    res.send(stringifyHealthCheck({ status: 'ok' }));
  });

  // app.register(fastifyWebsocket, {
  //   options: {
  //     maxPayload: 1048576,
  //     verifyClient: (info, next) => next(true),
  //   },
  // });

  app.register(MercuriusGQLUpload, {
    maxFileSize: 1024 * 1024 * 20,
    maxFiles: 1,
  });

  app.register(mercurius, {
    schema,
    context:
      contextFn ?? ((request): Context => new Context(request, connection)),
    queryDepth: 10,
    subscription: getSubscriptionSettings(connection),
    // Disable GraphQL introspection in production
    graphiql: !isProd,
    validationRules: isProd && [NoSchemaIntrospectionCustomRule],
    errorFormatter(execution, ctx) {
      if (execution.errors?.length > 0) {
        const flatErrors = execution.errors.flatMap<GraphQLError>((error) => {
          if (error.originalError.name === 'FastifyError') {
            return (
              (error.originalError as MercuriusError<GraphQLError>).errors ||
              error
            );
          }
          return error;
        });
        return {
          statusCode: 200,
          response: {
            data: execution.data,
            errors: flatErrors.map((error): GraphQLError => {
              const newError = error as Mutable<GraphQLError>;
              if (!error.originalError) {
                newError.extensions = {
                  code: 'GRAPHQL_VALIDATION_FAILED',
                };
              } else if (error.originalError?.name === 'EntityNotFoundError') {
                newError.message = 'Entity not found';
                newError.extensions = {
                  code: 'NOT_FOUND',
                };
              } else if (
                (error.originalError as FastifyError)?.code ===
                'MER_ERR_GQL_PERSISTED_QUERY_NOT_FOUND'
              ) {
                app.log.debug(
                  { body: ctx?.reply?.request?.body },
                  'unknown query',
                );
              } else if (!error.extensions?.code) {
                app.log.warn(
                  {
                    err: error.originalError,
                    body: ctx?.reply?.request?.body,
                  },
                  'unexpected graphql error',
                );
                newError.message = 'Unexpected error';
                newError.extensions = {
                  code: 'UNEXPECTED',
                };
              }
              if (isProd) {
                newError.originalError = undefined;
              }
              return newError;
            }),
          },
        };
      }
      return {
        statusCode: 200,
        response: execution,
      };
    },
  });

  if (isProd) {
    app.register(MercuriusCache, {
      ttl: 10,
      policy: {
        Query: {
          searchBookmarksSuggestions: {
            extendKey: userExtendKey,
          },
          searchBookmarks: {
            extendKey: userExtendKey,
          },
          // anonymousFeed: {
          //   extendKey: userExtendKey,
          // },
          // feed: {
          //   extendKey: userExtendKey,
          // },
          sourceFeed: {
            extendKey: userExtendKey,
          },
          tagFeed: {
            extendKey: userExtendKey,
          },
          keywordFeed: {
            extendKey: userExtendKey,
          },
          searchPostSuggestions: {
            extendKey: userExtendKey,
          },
          searchPosts: {
            extendKey: userExtendKey,
          },
          authorFeed: {
            extendKey: userExtendKey,
          },
          mostUpvotedFeed: {
            extendKey: userExtendKey,
          },
          mostDiscussedFeed: {
            extendKey: userExtendKey,
          },
          randomTrendingPosts: {
            extendKey: userExtendKey,
          },
          randomSimilarPostsByTags: {
            extendKey: userExtendKey,
          },
          randomDiscussedPosts: {
            extendKey: userExtendKey,
          },
          tagsCategories: true,
          advancedSettings: {
            extendKey: trackingExtendKey,
          },
          banner: true,
          postByUrl: {
            extendKey: userExtendKey,
          },
          post: {
            extendKey: userExtendKey,
          },
          postUpvotes: true,
          searchTags: true,
          userReadingRankHistory: true,
          userReadHistory: true,
        },
      },
    });
  }

  app.register(compatibility, { prefix: '/v1' });
  app.register(routes, { prefix: '/' });

  return app;
}
