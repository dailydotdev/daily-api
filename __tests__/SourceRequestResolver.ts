import { ApolloServer } from 'apollo-server-fastify';
import {
  ApolloServerTestClient,
  createTestClient,
} from 'apollo-server-testing';
import { Connection, getConnection } from 'typeorm';
import * as nock from 'nock';
import { FastifyInstance } from 'fastify';
import * as request from 'supertest';
import * as _ from 'lodash';

import { Context } from '../src/Context';
import createApolloServer from '../src/apollo';
import { authorizeRequest, MockContext } from './helpers';
import appFunc from '../src';
import { RequestSourceInput } from '../src/resolver';
import { Roles } from '../src/authChecker';
import { SourceRequest } from '../src/entity';
import { sourceRequestFixture } from './fixture/sourceRequest';

let con: Connection;
let server: ApolloServer;
let client: ApolloServerTestClient;
let loggedUser: string = null;

const mockInfo = (): nock.Scope =>
  nock(process.env.GATEWAY_URL)
    .get('/v1/users/me/info')
    .matchHeader('authorization', `Service ${process.env.GATEWAY_SECRET}`)
    .matchHeader('user-id', '1')
    .matchHeader('logged-in', 'true')
    .reply(200, { email: 'ido@daily.dev', name: 'Ido' });

const mockRoles = (roles: Roles[] = []): nock.Scope =>
  nock(process.env.GATEWAY_URL)
    .get('/v1/users/me/roles')
    .matchHeader('authorization', `Service ${process.env.GATEWAY_SECRET}`)
    .matchHeader('user-id', '1')
    .matchHeader('logged-in', 'true')
    .reply(200, roles);

beforeAll(async () => {
  con = await getConnection();
  server = await createApolloServer({
    context: (): Context => new MockContext(con, loggedUser),
    playground: false,
  });
  client = createTestClient(server);
});

beforeEach(async () => {
  loggedUser = null;
});

describe('mutation requestSource', () => {
  const MUTATION = `
  mutation RequestSource($data: RequestSourceInput!) {
  requestSource(data: $data) {
    sourceUrl
    userId
    userName
    userEmail
  }
}`;

  it('should not authorize when not logged in', async () => {
    const data: RequestSourceInput = { sourceUrl: 'http://source.com' };
    const res = await client.mutate({
      mutation: MUTATION,
      variables: { data },
    });
    expect(res.errors).toMatchSnapshot();
  });

  it('should return bad request when url is not valid', async () => {
    loggedUser = '1';
    const data: RequestSourceInput = { sourceUrl: 'invalid' };
    const res = await client.mutate({
      mutation: MUTATION,
      variables: { data },
    });
    expect(res.errors).toMatchSnapshot();
  });

  it('should add new source request', async () => {
    mockInfo();
    loggedUser = '1';
    const data: RequestSourceInput = { sourceUrl: 'http://source.com' };
    const res = await client.mutate({
      mutation: MUTATION,
      variables: { data },
    });
    expect(res.data).toMatchSnapshot();
  });
});

describe('query pendingSourceRequests', () => {
  const QUERY = (first = 10): string => `{
  pendingSourceRequests(first: ${first}) {
    pageInfo {
      endCursor
      hasNextPage
    }
    edges {
      node {
        sourceUrl
      }
    }
  }
}`;

  it('should not authorize when not moderator', async () => {
    mockRoles();
    loggedUser = '1';
    const res = await client.query({ query: QUERY() });
    expect(res.errors).toMatchSnapshot();
  });

  it('should return pending source requests', async () => {
    mockRoles([Roles.Moderator]);
    loggedUser = '1';

    for (const req of sourceRequestFixture) {
      await con.getRepository(SourceRequest).save(req);
    }

    const res = await client.query({ query: QUERY() });
    expect(res.data).toMatchSnapshot();
  });
});

describe('compatibility routes', () => {
  let app: FastifyInstance;

  beforeAll(async () => {
    app = await appFunc();
    return app.ready();
  });

  afterAll(() => app.close());

  describe('POST /publications/request', () => {
    it('should not authorize when not logged in', () => {
      return request(app.server)
        .post('/v1/publications/request')
        .send({ url: 'http://source.com' })
        .expect(401);
    });

    it('should return bad request when url is not valid', () => {
      return authorizeRequest(
        request(app.server).post('/v1/publications/request'),
      )
        .send({ url: 'invalid' })
        .expect(400);
    });

    it('should request new source', () => {
      mockInfo();
      return authorizeRequest(
        request(app.server).post('/v1/publications/request'),
      )
        .send({ url: 'http://source.com' })
        .expect(204);
    });

    it('should request new source (/requests)', () => {
      mockInfo();
      return authorizeRequest(
        request(app.server).post('/v1/publications/requests'),
      )
        .send({ url: 'http://source.com' })
        .expect(204);
    });
  });

  describe('GET /publications/requests/open', () => {
    it('should return pending source requests', async () => {
      mockRoles([Roles.Moderator]);

      for (const req of sourceRequestFixture) {
        await con.getRepository(SourceRequest).save(req);
      }

      const res = await authorizeRequest(
        request(app.server).get('/v1/publications/requests/open'),
      ).expect(200);
      const actual = res.body.map((x) => _.omit(x, ['id', 'createdAt']));
      expect(actual).toMatchSnapshot();
    });
  });
});
