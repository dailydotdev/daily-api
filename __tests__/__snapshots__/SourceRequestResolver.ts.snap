// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compatibility routes GET /publications/requests/open should return pending source requests 1`] = `
Array [
  Object {
    "approved": null,
    "closed": false,
    "pubId": null,
    "pubImage": null,
    "pubName": null,
    "pubRss": null,
    "pubTwitter": null,
    "url": "http://1.com",
    "userEmail": null,
    "userId": "1",
    "userName": null,
  },
  Object {
    "approved": true,
    "closed": false,
    "pubId": "a",
    "pubImage": "http://a.com",
    "pubName": "A",
    "pubRss": "http://a.com/feed",
    "pubTwitter": "a",
    "url": "http://3.com",
    "userEmail": null,
    "userId": "2",
    "userName": null,
  },
]
`;

exports[`mutation requestSource should add new source request 1`] = `
Object {
  "requestSource": Object {
    "sourceUrl": "http://source.com",
    "userEmail": "ido@daily.dev",
    "userId": "1",
    "userName": "Ido",
  },
}
`;

exports[`mutation requestSource should not authorize when not logged in 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "UNAUTHORIZED_ERROR",
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 3,
      },
    ],
    "message": "Access denied! You need to be authorized to perform this action!",
    "path": Array [
      "requestSource",
    ],
  },
]
`;

exports[`mutation requestSource should return bad request when url is not valid 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "VALIDATION_ERROR",
      "exception": Object {
        "validationErrors": Array [
          ValidationError {
            "children": Array [],
            "constraints": Object {
              "isUrl": "Must be a valid URL",
            },
            "property": "sourceUrl",
            "target": RequestSourceInput {
              "sourceUrl": "invalid",
            },
            "value": "invalid",
          },
        ],
      },
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 3,
      },
    ],
    "message": "Argument Validation Error",
    "path": Array [
      "requestSource",
    ],
  },
]
`;

exports[`query pendingSourceRequests should not authorize when not moderator 1`] = `
Array [
  Object {
    "extensions": Object {
      "code": "UNAUTHORIZED_ERROR",
    },
    "locations": Array [
      Object {
        "column": 3,
        "line": 2,
      },
    ],
    "message": "Access denied! You don't have permission for this action!",
    "path": Array [
      "pendingSourceRequests",
    ],
  },
]
`;

exports[`query pendingSourceRequests should return pending source requests 1`] = `
Object {
  "pendingSourceRequests": Object {
    "edges": Array [
      Object {
        "node": Object {
          "sourceUrl": "http://1.com",
        },
      },
      Object {
        "node": Object {
          "sourceUrl": "http://3.com",
        },
      },
    ],
    "pageInfo": Object {
      "endCursor": "YXJyYXljb25uZWN0aW9uOjE=",
      "hasNextPage": false,
    },
  },
}
`;
